File: services/ansible-service/internal/metrics/prometheus.py
"""Métriques Prometheus pour le service Ansible."""
import logging
from prometheus_client import Counter, Histogram, Gauge, generate_latest, CONTENT_TYPE_LATEST
from fastapi import Response

logger = logging.getLogger(__name__)

# Compteurs
api_requests_total = Counter(
    "ansible_api_requests_total",
    "Nombre total de requêtes API",
    ["method", "endpoint", "status"],
)

tower_api_requests_total = Counter(
    "ansible_tower_api_requests_total",
    "Nombre total de requêtes vers Ansible Tower",
    ["method", "endpoint", "status"],
)

jobs_launched_total = Counter(
    "ansible_jobs_launched_total",
    "Nombre total de jobs lancés",
    ["template_id"],
)

webhooks_received_total = Counter(
    "ansible_webhooks_received_total",
    "Nombre total de webhooks reçus",
    ["event_type"],
)

# Histogrammes
api_request_duration_seconds = Histogram(
    "ansible_api_request_duration_seconds",
    "Durée des requêtes API en secondes",
    ["method", "endpoint"],
    buckets=[0.1, 0.5, 1.0, 2.0, 5.0, 10.0],
)

tower_api_request_duration_seconds = Histogram(
    "ansible_tower_api_request_duration_seconds",
    "Durée des requêtes vers Ansible Tower en secondes",
    ["method", "endpoint"],
    buckets=[0.1, 0.5, 1.0, 2.0, 5.0, 10.0],
)

# Jauges
active_websocket_connections = Gauge(
    "ansible_active_websocket_connections",
    "Nombre de connexions WebSocket actives",
)

cache_hits_total = Counter(
    "ansible_cache_hits_total",
    "Nombre total de hits de cache",
    ["cache_type"],
)

cache_misses_total = Counter(
    "ansible_cache_misses_total",
    "Nombre total de misses de cache",
    ["cache_type"],
)


def get_metrics_response() -> Response:
    """Retourne la réponse avec les métriques Prometheus."""
    return Response(
        content=generate_latest(),
        media_type=CONTENT_TYPE_LATEST,
    )

File: services/pipeline-service/internal/metrics/prometheus.go
package metrics

import (
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/promauto"
)

var (
	// WebhooksReceivedTotal compte les webhooks reçus par provider
	WebhooksReceivedTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "pipeline_webhooks_received_total",
			Help: "Nombre total de webhooks reçus",
		},
		[]string{"provider", "status"},
	)

	// WebhookProcessingDuration mesure la durée de traitement des webhooks
	WebhookProcessingDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "pipeline_webhook_processing_duration_seconds",
			Help:    "Durée de traitement des webhooks en secondes",
			Buckets: prometheus.ExponentialBuckets(0.001, 2, 12),
		},
		[]string{"provider"},
	)

	// RunsStoredTotal compte les runs stockés
	RunsStoredTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "pipeline_runs_stored_total",
			Help: "Nombre total de runs stockés",
		},
		[]string{"provider", "status"},
	)

	// APIRequestsTotal compte les requêtes API
	APIRequestsTotal = promauto.NewCounterVec(
		prometheus.CounterOpts{
			Name: "pipeline_api_requests_total",
			Help: "Nombre total de requêtes API",
		},
		[]string{"method", "endpoint", "status"},
	)

	// APIRequestDuration mesure la durée des requêtes API
	APIRequestDuration = promauto.NewHistogramVec(
		prometheus.HistogramOpts{
			Name:    "pipeline_api_request_duration_seconds",
			Help:    "Durée des requêtes API en secondes",
			Buckets: []float64{0.01, 0.05, 0.1, 0.25, 0.5, 1, 2.5, 5},
		},
		[]string{"method", "endpoint"},
	)
)
